### PR TITLE
Add CometLogger for metrics logging via comet.ml

### PR DESCRIPTION
**Pull request by dzheng256**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

My company uses torchtune for LLM finetuning but we use comet.ml for experiment tracking, not tensorboard or weights and biases. Adding a `CometLogger` class to make it easier for other Comet users to use torchtune with Comet.

#### Changelog
- Introduce `CometLogger` class
- Update docs to include `CometLogger` 
- Add `TestCometLogger` test

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
	- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

**Comment by joecummings** [comment]
@dzheng256  Would you mind signing the CLA agreement?

**Comment by dzheng256** [comment]
Hey @joecummings I actually already had an email exchange with the meta CLA folks about this and I think I'm on the agreement already but it might be my corporate email (danz@netflix.com) rather than the account I'm making this PR from (dzheng256@gmail.com). I've asked them to add my personal email to the corporate CLA as well but haven't heard back.

**Comment by dzheng256** [comment]
It seems like the tests ran and passed, @joecummings does that mean the CLA issue is resolved?

**Comment by RdoubleA** [comment]
@dzheng256 It seems the CLA check is still failing. Have you confirmed that your personal email was added? Are you still able to sign a second time with your personal email?

**Comment by dzheng256** [comment]
@joecummings sorry about that, seems like it's all resolved now.

**Comment by SalmanMohammadi** [comment]
Thank you so much for adding this @dzheng256. It's really cool to hear that you're using torchtune in your company.

Could you possibly attach an example comet logging output to the PR to confirm everything looks GOOD?

**Comment by SalmanMohammadi** [comment]
Could you also please shout out integration in the README.md here, please? https://github.com/pytorch/torchtune/blob/288ff4435b0cf17325b5c3b112f6859a6cdf0ea2/README.md?plain=1#L35

**Comment by RdoubleA** [comment]
Thanks for adding this. There was also an earlier PR to add CometLogger that including a quick tutorial on how to use it. Would you be able to add that in this PR as well? #910

**Comment by dzheng256** [comment]
Oh sorry, I hadn't realized there was an open PR implementing pretty much the same thing as this one. Should we stick with #910 then, since it came first? It also seems to be written by the a Comet dev so it might provide a better implementation. cc @Lothiraldan

**Comment by RdoubleA** [comment]
@dzheng256 I've pinged the other PR to see if he's still willing to work on it. If not, maybe you can incorporate the differences here?

**Comment by dzheng256** [comment]
Yes sounds good, I will pull in the tutorial from the other PR if they no longer have the bandwidth to work on it.

File: pyproject.toml
Comment (Lothiraldan) [inline pyproject.toml:45]:
```suggestion
    "comet_ml>=3.44.2",
```
This version contains our latest change to our new way of initializing Comet Experiment

File: torchtune/utils/metric_logging.py
Comment (Lothiraldan) [inline torchtune/utils/metric_logging.py:376]: We are moving towards a new way of initializing Comet Experiment with a new top-level function: `comet_ml.start`. This has several benefits:
* It support online and offline logging
* It support logging a new experiment or appending to an existing Experiment
* You can set name and tags directly at initialization time

I'm working right now on finalizing the documentation but I can share the reference doc for that new function already: https://www.comet.com/docs/v2/api-and-sdk/python-sdk/reference/Experiment-Creation/

If there is a chance that users might want to append data to an existing experiment or create log data to an offline Experiment (to upload later), I would recommend using that instead. Please let me know, happy to provide the code corresponding code change.

File: torchtune/utils/metric_logging.py
Comment (Lothiraldan) [inline torchtune/utils/metric_logging.py:323]:
```suggestion
    """Logger for use w/ Comet (https://www.comet.com/site/).
```
We changed our domain some time ago

**Review by Lothiraldan — COMMENTED** [review]
Looking good so far, I will give it a try in the next days to generate an example public project.

@dzheng256 Feel free to pull the tutorial from my PR https://github.com/pytorch/torchtune/pull/910/files#diff-5c3c14a3a87e5106e63f046499c113cd5b5c0771ac548bcb5b097fc40342bd44, I don't think I have the permission to push to your fork. Once I have a new example project, I will share the links to update the tutorial

There is one open question, whether to use our new way of initializing Comet Experiments. This is a new way but it's about to be more globally released, we have it stable for a while now and we already updated several of our integrations to use it, last one was Transformers: https://github.com/huggingface/transformers/pull/31366

File: torchtune/utils/metric_logging.py
Comment (dzheng256) [reply torchtune/utils/metric_logging.py]: Thanks @Lothiraldan, I swapped it over to the new comet_ml.start() and verified the latest code works for me. Could you use it to make an example project which I can then link in the deep dive tutorial?

File: torchtune/utils/metric_logging.py
Comment (dzheng256) [reply torchtune/utils/metric_logging.py]: done

**Comment by dzheng256** [comment]
@Lothiraldan There are still a few differences compared to your original PR, specifically in the initialization args as well as the log metric calls. I noticed you used `self._experiment.__internal_api__log_metric__` for example. Let me know if you feel we should tweak anything.

File: torchtune/utils/metric_logging.py
Comment (dzheng256) [reply torchtune/utils/metric_logging.py]: @Lothiraldan I also added you as a collaborator to my fork, feel free to directly push changes if you would like

File: docs/source/deep_dives/comet_logging.rst
Comment (SalmanMohammadi) [inline docs/source/deep_dives/comet_logging.rst:20]: I think you need to grab this png from the original PR, unless it needs to be updated @Lothiraldan?

File: docs/source/deep_dives/comet_logging.rst
Comment (SalmanMohammadi) [inline docs/source/deep_dives/comet_logging.rst:49]: Looks like this is the original link with a run back in April? @ dzheng256 could you please demonstrate everything works okay by adding an example of a comet run using this PR?

@Lothiraldan feel free to chime in if you'd rather this run be on an official comet channel

File: pyproject.toml
Comment (SalmanMohammadi) [inline pyproject.toml:45]: to confirm, this is the correct version for the comet v2 API?

File: docs/source/deep_dives/comet_logging.rst
Comment (Lothiraldan) [reply docs/source/deep_dives/comet_logging.rst:20]: I will regenerate a new screenshot after re-running the example with the latest version of this PR, that's on me

File: docs/source/deep_dives/comet_logging.rst
Comment (Lothiraldan) [reply docs/source/deep_dives/comet_logging.rst:49]: Happy to help, one benefit on having the example in our "officlal" workspace is that we can more easily update it when we make change to integrations so it stay up-to-date. But I don't have a strong opinion about it.

To be fully transparent, I was hoping to just restart the previous EC2 instance I used back in April but I am running into some quota issue. If @dzheng256 have more easy access to a training machine and is happy to run an experiment, we could get the PR merged more quickly. Again happy to go with whatever you recommend

File: pyproject.toml
Comment (Lothiraldan) [reply pyproject.toml:45]: Yes, this version contains our latest change and fixes related to `comet_ml.start` and relevant to this integration

**Comment by Lothiraldan** [comment]
> @Lothiraldan There are still a few differences compared to your original PR, specifically in the initialization args as well as the log metric calls. I noticed you used `self._experiment.__internal_api__log_metric__` for example. Let me know if you feel we should tweak anything.

We are generally using this integration specific methods to have more control how the data is logged. I don't think this makes a different for the torchtune integration but I will double-check once we get the example running. Thanks for flagging that.

**Comment by SalmanMohammadi** [comment]
This is looking great to me. Thank you both so much for your help @dzheng256 @Lothiraldan. I'm really happy to see torchtune being integrated into the wider ML ecosystem.

Could you ensure your pre-commit hooks are correctly configured, please? Looks like the lint CI is failing. 

RE: example runs, I don't want you to be blocked on this, so I'm happy to help out here in any way I can. 

I'd also love to see our integration demonstrated through your "official" channel : ). We'd very much welcome future PRs to ensure users can access the latest and greatest features in Comet through torchtune.

File: docs/source/deep_dives/comet_logging.rst
Comment (SalmanMohammadi) [inline docs/source/deep_dives/comet_logging.rst:31]: I think you need a blank line here

File: docs/source/deep_dives/comet_logging.rst
Comment (SalmanMohammadi) [inline docs/source/deep_dives/comet_logging.rst:38]: and here

File: docs/source/deep_dives/comet_logging.rst
Comment (SalmanMohammadi) [reply docs/source/deep_dives/comet_logging.rst:31]:
```suggestion

    comet init --api-key
```

File: docs/source/deep_dives/comet_logging.rst
Comment (SalmanMohammadi) [reply docs/source/deep_dives/comet_logging.rst:38]:
```suggestion

    # enable logging to the built-in CometLogger
```

File: docs/source/deep_dives/comet_logging.rst
Comment (SalmanMohammadi) [inline docs/source/deep_dives/comet_logging.rst:43]:
```suggestion
      experiment_name: my-experiment-name
      
```

File: torchtune/utils/metric_logging.py
Comment (SalmanMohammadi) [inline torchtune/utils/metric_logging.py:331]:
```suggestion
```

File: torchtune/utils/metric_logging.py
Comment (SalmanMohammadi) [inline torchtune/utils/metric_logging.py:339]:
```suggestion
        mode (Optional[str]): Control how the Comet experiment is started. "get": Continue logging to an existing
            experiment identified by the `experiment_key` value. "create": Always creates of a new experiment, useful
            for HPO sweeps. "get_or_create" (default): Starts a fresh experiment if required, or persists logging to
            an existing one.
        online (Optional[bool]): If True, the data will be logged to Comet server, otherwise it will be stored locally
            in offline experiment. Default is `True`.
```

nit: ordering

Comment (Lothiraldan) [reply torchtune/utils/metric_logging.py]: Thanks, I was wondering why the linting was failing initially.

File: docs/source/deep_dives/comet_logging.rst
Comment (RdoubleA) [inline docs/source/deep_dives/comet_logging.rst:17]: nit: we lowercase torchtune everywhere, even at the start of a sentence :)

**Comment by Lothiraldan** [comment]
@SalmanMohammadi I got back access to the EC2 instance and I've got the example running: https://www.comet.com/examples/comet-example-torchtune-mistral/fffb3036880e41b5af2df932db4d3578?compareXAxis=step&experiment-tab=panels&showOutliers=true&smoothing=0&xAxis=step. I took a screenshot while it's running, it will probably looks better after it finishes but we can merge with the current screenshot.

I made some small changes and added back logging the configuration as a file, this way it's pretty easy to grab it from Comet and reproduce a previous run. I made several small commits, sorry about that. Let me know if you want me to squash them if you are not gonna squash the PR.

File: docs/source/deep_dives/comet_logging.rst
Comment (RdoubleA) [inline docs/source/deep_dives/comet_logging.rst:28]: did you intend to add an example pip command?

File: docs/source/deep_dives/comet_logging.rst
Comment (RdoubleA) [inline docs/source/deep_dives/comet_logging.rst:43]: 😍

File: docs/source/deep_dives/comet_logging.rst
Comment (RdoubleA) [inline docs/source/deep_dives/comet_logging.rst:52]:
```suggestion
  Click on this sample `Comet project to see the logged metrics after fine-tuning <https://www.comet.com/examples/comet-example-torchtune-mistral/>`_.
```

File: torchtune/utils/metric_logging.py
Comment (RdoubleA) [inline torchtune/utils/metric_logging.py:323]: Could you add a quick sentence describing what Comet is? (I know we didn't have that for the WandBLogger)

File: torchtune/utils/metric_logging.py
Comment (RdoubleA) [inline torchtune/utils/metric_logging.py:328]: should we just encourage users to use `comet login`?

File: torchtune/utils/metric_logging.py
Comment (RdoubleA) [inline torchtune/utils/metric_logging.py:332]: what is this needed for?

File: torchtune/utils/metric_logging.py
Comment (RdoubleA) [inline torchtune/utils/metric_logging.py:334]: you can make a bulleted list here
```suggestion
        mode (Optional[str]): Control how the Comet experiment is started. 
        
        - ``"get"``: Continue logging to an existing [...]
        - ``"create"``: [...]
        - ``"get_or_create"``: [...]
```
also minor nit, can we make the default option the first bullet?

File: torchtune/utils/metric_logging.py
Comment (RdoubleA) [inline torchtune/utils/metric_logging.py:335]:
```suggestion
            experiment identified by the ``experiment_key`` value. "create": Always creates of a new experiment, useful
```

File: torchtune/utils/metric_logging.py
Comment (RdoubleA) [inline torchtune/utils/metric_logging.py:339]:
```suggestion
            in an offline experiment. Default is ``True``.
```

File: torchtune/utils/metric_logging.py
Comment (RdoubleA) [inline torchtune/utils/metric_logging.py:343]:
```suggestion
        **kwargs: additional arguments to pass to ``Comet.start``. See
```

File: torchtune/utils/metric_logging.py
Comment (RdoubleA) [inline torchtune/utils/metric_logging.py:354]:
```suggestion
        ImportError: If ``comet_ml`` package is not installed.
```

**Comment by SalmanMohammadi** [comment]
> @SalmanMohammadi I got back access to the EC2 instance and I've got the example running: https://www.comet.com/examples/comet-example-torchtune-mistral/fffb3036880e41b5af2df932db4d3578?compareXAxis=step&experiment-tab=panels&showOutliers=true&smoothing=0&xAxis=step. I took a screenshot while it's running, it will probably looks better after it finishes but we can merge with the current screenshot.
> 
> I made some small changes and added back logging the configuration as a file, this way it's pretty easy to grab it from Comet and reproduce a previous run. I made several small commits, sorry about that. Let me know if you want me to squash them if you are not gonna squash the PR.

Amazing!! Thank you. 

Feel free to send me an updated screenshot on Discord and I'm happy to add.

File: torchtune/utils/metric_logging.py
Comment (RdoubleA) [inline torchtune/utils/metric_logging.py:361]: personally I find comet login the easiest to do and is more comfortable for folks who don't know much code. What are your thoughts on just always recommending to setup via comet login so folks do not get confused with the different options?

File: torchtune/utils/metric_logging.py
Comment (SalmanMohammadi) [reply torchtune/utils/metric_logging.py:331]: I think it's still unhappy about this @Lothiraldan

File: torchtune/utils/metric_logging.py
Comment (SalmanMohammadi) [inline torchtune/utils/metric_logging.py:343]: might need to type this (annoyingly)
```suggestion
        **kwargs: additional arguments to pass to Comet.start. See
```

File: torchtune/utils/metric_logging.py
Comment (SalmanMohammadi) [inline torchtune/utils/metric_logging.py:375]: and here

**Review by RdoubleA — COMMENTED** [review]
Looks great overall. I have a lot of nit comments around RST docs rendering stuff, hope you don't mind :)

File: torchtune/utils/metric_logging.py
Comment (Lothiraldan) [reply torchtune/utils/metric_logging.py:331]: Yes, I'm working on understanding why is it complaining. It only is complaining with the latest version of pydocstring.

File: torchtune/utils/metric_logging.py
Comment (sherpan) [reply torchtune/utils/metric_logging.py:323]: @Lothiraldan 

"Comet is an experiment tracking tool that helps ML teams track, debug, compare, and reproduce their model training runs"

File: torchtune/utils/metric_logging.py
Comment (Lothiraldan) [reply torchtune/utils/metric_logging.py:363]: `comet login` works great in interactive situations but won't help in automated setups. I'm happy to highlight the `comet login` option and links to other options in the Comet docs if that works for you.

File: torchtune/utils/metric_logging.py
Comment (Lothiraldan) [reply torchtune/utils/metric_logging.py:330]: Yes we should encourage users to use `comet login`.

Saving your Comet API Key in a git tracked file is a bad practice and we are of course not recommending it. Having the possibility to set your Comet API Key in your configuration is useful when you are rendering your configuration file and injecting the Comet API Key automatically from a secret manager place.

I do not know if that's something that's frequent when using torchtune.

File: torchtune/utils/metric_logging.py
Comment (Lothiraldan) [reply torchtune/utils/metric_logging.py:334]: The `experiment_key` is used in two situations:

* When resuming a Comet Experiment, it is the ID of the Experiment to resume logging data to.
* When creating a new Comet Experiment, we generate a random ID by default. In various cases, it is useful to control the new ID (like having the same ID for the Comet Experiment and the Cloud provider training job).

File: torchtune/utils/metric_logging.py
Comment (Lothiraldan) [reply torchtune/utils/metric_logging.py]: I give it a try but I cannot seems to have found the right format. Here is how it looks if I align it with the arguments list 
Screenshot 2024-08-08 at 21-36-06 CometLogger — torchtune main documentation

And I indent them even by one space, sphinx-build is complaining:

```
.../torchtune/torchtune/utils/metric_logging.py:docstring of torchtune.utils.metric_logging.CometLogger:20:Unexpected indentation.
```

Do you know what is the proper way to format it?

File: docs/source/deep_dives/comet_logging.rst
Comment (Lothiraldan) [reply docs/source/deep_dives/comet_logging.rst:28]: Good point, added it

File: docs/source/deep_dives/comet_logging.rst
Comment (dzheng256) [reply docs/source/deep_dives/comet_logging.rst]: I have access to machines but only for netflix-internal inrfa. If you can set me up with the authentication for logging to the official workspace I could do that, but otherwise my runs are logged to our internal comet instance.

File: torchtune/utils/metric_logging.py
Comment (dzheng256) [reply torchtune/utils/metric_logging.py:331]: Actually I expect people will mostly just use the API key? By exporting it before running `tune run ...`. Otherwise, where would they call `comet.login()`?

File: torchtune/utils/metric_logging.py
Comment (dzheng256) [reply torchtune/utils/metric_logging.py:364]: Same comment as in https://github.com/pytorch/torchtune/pull/1221#discussion_r1710221128, if we expect users to be setting up a config and doing `tune run ...` how are they supposed to use `comet.login()`? We would have to call that inside the logger, and incorporate some API key via a secrets manager solution which would be different for every user.

As far as I can tell torchtune seems to heavily favor the CLI `tune run` style usage, rather than importing it and calling tune run programmatically. If this is indeed the case, I think we should actually remove the mention to `comet.login()` and have people export the api key, what do you guys think?

Comment (RdoubleA) [reply torchtune/utils/metric_logging.py:364]: Sorry I was under the impression you can run `comet login` from CLI, is this not the case?

Either way I agree with @dzheng256, since CLI is the primary entry point for most users via `tune run` I would lean towards suggesting whatever fits that workflow best whether that's setting environment variables or a CLI command

Comment (Lothiraldan) [reply torchtune/utils/metric_logging.py:364]: @RdoubleA @dzheng256 Sorry about the confusion, there is bot a `comet login` command-line script and `comet_ml.login`  that one can call in their Python script. I will clean up and double-check everything to point people to use the CLI `comet login`.

File: torchtune/utils/metric_logging.py
Comment (Lothiraldan) [reply torchtune/utils/metric_logging.py]: I re-ordered the options but couldn't find the right format to render an indented list. I'm not sure if the Google docstring support it.

File: tests/torchtune/utils/test_metric_logging.py
Comment (SalmanMohammadi) [inline tests/torchtune/utils/test_metric_logging.py:187]:
```suggestion
            logger = CometLogger(project="test_project")
```

File: tests/torchtune/utils/test_metric_logging.py
Comment (SalmanMohammadi) [inline tests/torchtune/utils/test_metric_logging.py]:
```suggestion
            logger = CometLogger(project="test_project")
```

**Comment by Lothiraldan** [comment]
Hi @SalmanMohammadi, I've pushed fixes for the unit tests, updated the links and screenshots, tried to clarify the docstring about `experiment_key` argument and found the correct format for the `mode` argument to render as a nested bullet point.

I think that we are all good now but please let me know if I missed anything

**Comment by SalmanMohammadi** [comment]
Thank you so much for your patience in addressing all our comments, and for your contributions @dzheng256, @Lothiraldan. This looks fantastic to me. 

One last tiny nit and feel free to merge once CI is green.

**Review by SalmanMohammadi — APPROVED** [review]

**Comment by Lothiraldan** [comment]
@SalmanMohammadi Good catch, thanks you and @RdoubleA  for all of your time and reviews. I hope that my millions of small commits were not spamming your inbox too much :sweat_smile:

**Comment by dzheng256** [comment]
Thanks everyone! Looking forward to deleting my internal CometLogger and switching to `torchtune.utils.metric_logging.CometLogger`

**Pull request merged**
